Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/cache/downloader/PartialContentSupportChecker.kt

```diff
@@ -0,0 +1,288 @@
+package com.github.adamantcheese.chan.core.cache.downloader
+
+import android.util.LruCache
+import androidx.annotation.GuardedBy
+import com.github.adamantcheese.chan.core.cache.FileCacheV2
+import com.github.adamantcheese.chan.core.cache.downloader.DownloaderUtils.isCancellationError
+import com.github.adamantcheese.chan.core.di.NetModule
+import com.github.adamantcheese.chan.utils.Logger
+import io.reactivex.Single
+import io.reactivex.SingleEmitter
+import okhttp3.*
+import okhttp3.HttpUrl.Companion.toHttpUrlOrNull
+import java.io.IOException
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.TimeoutException
+
+/**
+ * This class is used to figure out whether an image or a file can be downloaded from the server in
+ * separate chunks concurrently using HTTP Partial-Content. For batched image downloading and
+ * media prefetching this method returns false because we should download them normally. Chunked
+ * downloading should only be used for high priority files/images like in the gallery when the user
+ * is viewing them. Everything else should be downloaded in a singe chunk.
+ * */
+internal class PartialContentSupportChecker(
+        private val okHttpClient: OkHttpClient,
+        private val activeDownloads: ActiveDownloads,
+        private val maxTimeoutMs: Long
+) {
+    // Thread safe
+    private val cachedResults = LruCache<String, PartialContentCheckResult>(1024)
+    @GuardedBy("itself")
+    private val checkedChanHosts = mutableMapOf<String, Boolean>()
+    private val badSites = setOf(
```

Comment:
I dunno if this is the best way of dealing with "bad" sites.

Reply:
2ch.hk sends file size in KBs instead of bytes which makes it impossible to calculate the exact file size which breaks chunking. But yeah, maybe it should be in resources or something?

Reply:
And it probably should be called something like "unsupported sites"? Or, "sites that send incorrect file size"?

Reply:
In any case, this doesn't mean that the whole concurrent chunked download won't work on 2ch.hk and other "bad sites". It just means that we can't trust the file size we get along with thread json from those sites and need to send HEAD request to such sites every time. So basically this optimisation won't work on those site, we will have to send HEAD request every time.

Reply:
I think I came up with an idea of where I should store this info. In classes that implemente Site interface. Then I can just inject SiteRepository, and with it's help check if the current site is "bad".

Reply:
Gonna do this one tomorrow too.

Reply:
Damn, this one will take more time than I expected.

Reply:
Done.